source/event_stream.c

```diff
@@ -0,0 +1,1118 @@
+                              "insufficient buffer length", LIB_NAME),
+        AWS_DEFINE_ERROR_INFO(aws_error_prelude_checksum_failure, AWS_ERROR_EVENT_STREAM_PRELUDE_CHECKSUM_FAILURE,
+                              "prelude checksum was incorrect", LIB_NAME),
+        AWS_DEFINE_ERROR_INFO(aws_error_message_checksum_failure, AWS_ERROR_EVENT_STREAM_MESSAGE_CHECKSUM_FAILURE,
+                              "message checksum was incorrect", LIB_NAME),
+        AWS_DEFINE_ERROR_INFO(aws_error_message_invalid_headers_length,
+                              AWS_ERROR_EVENT_STREAM_MESSAGE_INVALID_HEADERS_LEN,
+                              "message headers length was incorrect", LIB_NAME),
+        AWS_DEFINE_ERROR_INFO(aws_error_unknown_header_type, AWS_ERROR_EVENT_STREAM_MESSAGE_UNKNOWN_HEADER_TYPE,
+                              "An unknown header type was encountered", LIB_NAME),
+        AWS_DEFINE_ERROR_INFO(aws_error_illegal_parser_state, AWS_ERROR_EVENT_STREAM_MESSAGE_PARSER_ILLEGAL_STATE,
+                              "message parser encountered an illegal state", LIB_NAME),
+};
+
+static struct aws_error_info_list list = {
+        .error_list = errors,
+        .count = sizeof(errors) / sizeof(struct aws_error_info),
+};
+
+static int8_t error_strings_loaded = 0;
+
+void aws_event_stream_load_error_strings(void) {
+    if (!error_strings_loaded) {
+        error_strings_loaded = 1;
+        aws_register_error_info(&list);
+    }
+}
+
+#define TOTAL_LEN_OFFSET 0
+#define PRELUDE_CRC_OFFSET (sizeof(uint32_t) + sizeof(uint32_t))
+#define HEADER_LEN_OFFSET sizeof(uint32_t)
+
+/* Computes the byte length necessary to store the headers represented in the headers list.
+ * returns that length. */
+uint32_t compute_headers_len(struct aws_array_list *headers) {
+    if (!headers || !aws_array_list_length(headers)) return 0;
+
+    size_t headers_count = aws_array_list_length(headers);
+    size_t headers_len = 0;
+
+    for (size_t i = 0; i < headers_count; ++i) {
+        struct aws_event_stream_header_value_pair *header;
+
+        aws_array_list_get_at_ptr(headers, (void **) &header, i);
+
+        headers_len += sizeof(header->header_name_len) +
+                       header->header_name_len + 1;
+
+        if (header->header_value_type == AWS_EVENT_STREAM_HEADER_STRING ||
+            header->header_value_type == AWS_EVENT_STREAM_HEADER_BYTE_BUF) {
+            headers_len += sizeof(header->header_value_len);
+        }
+
+        if (header->header_value_type != AWS_EVENT_STREAM_HEADER_BOOL_FALSE &&
+            header->header_value_type != AWS_EVENT_STREAM_HEADER_BOOL_TRUE) {
+            headers_len += header->header_value_len;
+        }
+    }
+
+    return (uint32_t) headers_len;
+}
+
+/* adds the headers represented in the headers list to the buffer.
+ returns the new buffer offset for use elsewhere. */
+size_t add_headers_to_buffer(struct aws_array_list *headers, uint8_t *buffer, size_t buffer_length) {
```

Comment:
`buffer_length` is not used and it should be.
Right now we can easily overflow the buffer since we're not checking when we're advancing `buffer_alias` if we went too far.